### PR TITLE
character detail UI added

### PR DESCRIPTION
**Pull request by benjaglobant**
In this PR I have implemented the character detail UI to show the details of a specific character selected from the list of characters shown in CharactersFragment. This UI was implemented using a collapsing toolbar and a nested scroll view.

Gif showing UI (Nexus 5X):
characterDetailUIPRNexus5X

Gif showing UI (Nexus S):
characterDetailUIPRNexusS

### CODE REVIEW

File: app/src/main/java/com/globant/harrypotterapp/adapter/CharactersAdapter.kt [hunk 1]
```diff
@@ -8,7 +8,11 @@ import com.globant.harrypotterapp.R
 import com.globant.domain.entity.Character
 import com.globant.harrypotterapp.databinding.FragmentCharactersCardViewBinding
 
-class CharactersAdapter : RecyclerView.Adapter<CharactersAdapter.ViewHolder>() {
+class CharactersAdapter(private val onCharacterClick: OnCharacterClick) : RecyclerView.Adapter<CharactersAdapter.ViewHolder>() {
+
+    interface OnCharacterClick {
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/adapter/CharactersAdapter.kt]: this interface should be outside the class

File: app/src/main/res/layout/activity_character_detail.xml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                    style="@style/style_character_detail_activity_title_text_view"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:text="@string/character_detail_activity_title_text" />
+
+            </LinearLayout>
+
+        </com.google.android.material.appbar.CollapsingToolbarLayout>
+
+        <androidx.appcompat.widget.Toolbar
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_character_detail.xml]: this should be inside the collapsing

**Review by SartiMau — CHANGES_REQUESTED** [review]
please pass character_detail_image to webp

File: app/src/main/java/com/globant/harrypotterapp/adapter/CharactersAdapter.kt (hunk 1 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/adapter/CharactersAdapter.kt]: You are right! Refactored!

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -12,8 +12,9 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/Theme.AppCompat.NoActionBar">
-        <activity android:name=".activity.SpellsActivity"></activity>
-        <activity android:name=".activity.HouseActivity"></activity>
+        <activity android:name=".activity.CharacterDetailActivity"></activity>
```
Comment (lucianoeli) [inline app/src/main/AndroidManifest.xml]: you can close the activity tag as you did below.

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        super.onCreate(savedInstanceState)
+        binding = ActivityCharacterDetailBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        characterDetailViewModel.getCharacterDetailLiveData().observe(::getLifecycle, ::updateUI)
+        characterDetailViewModel.fetchCharacterDetail()
+    }
+
+    private fun updateUI(data: CharacterDetail) {
+        binding.characterDetailActivityName.text = data.name
```
Comment (lucianoeli) [inline app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt]: you can use:
```
with(binding){
}
```

File: app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt [hunk 5]
```diff
@@ -58,7 +61,9 @@ class CharactersFragment : Fragment() {
         Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
     }
 
-    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? = binding.root
+    override fun showCharacterDetail() {
+        this.context?.let { startActivity(CharacterDetailActivity.getIntent(it)) }
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt:73]: this should call the viewModel to set the new state of this liveData.. then you will call the startActivity in the fragment

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            getString(R.string.character_detail_activity_order_of_the_phoenix_text, getStringValueForDetail(data.orderOfThePhoenix))
+        binding.characterDetailActivityDumbledoresArmy.text =
+            getString(R.string.character_detail_activity_dumbledores_army_text, getStringValueForDetail(data.dumbledoresArmy))
+        binding.characterDetailActivityDeathEater.text =
+            getString(R.string.character_detail_activity_death_eater_text, getStringValueForDetail(data.deathEater))
+        binding.characterDetailActivityBloodStatus.text =
+            getString(R.string.character_detail_activity_blood_status_text, data.bloodStatus)
+        binding.characterDetailActivitySpecies.text =
+            getString(R.string.character_detail_activity_species_text, data.species)
+    }
```
Comment (nicolbidone) [inline app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt:43]: her you can use with(binding)

**Comment by nicolbidone** [comment]
are all the images converted to webp?

File: app/src/main/res/layout/activity_character_detail.xml (hunk 2 above)
Comment (benjaglobant) [reply app/src/main/res/layout/activity_character_detail.xml]: Done!

File: app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (benjaglobant) [reply app/src/main/AndroidManifest.xml]: Done!

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt (hunk 4 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt]: Done!

File: app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt (hunk 5 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt:73]: Done!

**Review by lucianoeli — APPROVED** [review]

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        super.onCreate(savedInstanceState)
+        binding = ActivityCharacterDetailBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        characterDetailViewModel.getCharacterDetailLiveData().observe(::getLifecycle, ::updateUI)
+        characterDetailViewModel.fetchCharacterDetail()
+    }
+
+    private fun updateUI(data: CharacterDetail) {
+        with(binding) {
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt:38]: this is not fine, call a method to update the XML info. this method should have 
when (data.state) {
     state1 -> {..}
     state2 -> {..}
}

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            characterDetailActivitySpecies.text =
+                getString(R.string.character_detail_activity_species_text, data.species)
+        }
+    }
+
+    private fun getStringValueForDetail(detail: Boolean): String = if (detail) YES else NO
+
+    companion object {
+        fun getIntent(context: Context): Intent = Intent(context, CharacterDetailActivity::class.java)
+        private const val NO = "No"
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt]: this two show be in a string resource

File: app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt (hunk 5 above)
Comment (SartiMau) [reply app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt:73]: this still here

File: app/src/main/java/com/globant/harrypotterapp/viewmodel/CharactersViewModel.kt [hunk 9]
```diff
@@ -35,12 +35,17 @@ class CharactersViewModel(private val getCharactersUseCase: GetCharactersUseCase
             }
         }
     }
+
+    override fun onCharacterClicked() {
+        charactersMutableLiveData.postValue(Event(CharactersData(status = CharactersStatus.OPEN_CHARACTER_DETAIL)))
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/viewmodel/CharactersViewModel.kt]: you are not into a coroutine, so in this case, you should use charactersMutableLiveData.value =

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt (hunk 5 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt:73]: Sorry Mau, refactored!

File: app/src/main/java/com/globant/harrypotterapp/viewmodel/CharactersViewModel.kt (hunk 9 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/viewmodel/CharactersViewModel.kt]: You are right! Refactored!

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt (hunk 8 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt]: Refactored!

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt (hunk 7 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt:38]: Done Mau, thanks!

**Review by mauroLonghi — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by hernanhorquin96 — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Pull request merged**